vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java

```diff
@@ -114,17 +114,17 @@ public void selectFromClient(T item) {
             fireSelectionEvent(new MultiSelectionEvent<>(getGrid(),
                     getGrid().asMultiSelect(), oldSelection, true));
 
-            long size = 0;
-
             if (!isSelectAllCheckboxVisible()) {
                 // Skip changing the state of Select All checkbox if it was
                 // meant to be hidden
                 return;
             }
 
-            size = getDataProviderSize();
-
+            long size = getDataProviderSize();
             selectionColumn.setSelectAllCheckboxState(size == selected.size());
```

Comment:
It took me a while before I realized that `size === selected.size()` will never be true when a hierarchical data provided is used. This might an argument for adding the `!isHierarchicalDataProvider() &&`  part to this condition to make it more explicit, same as in [`doUpdateSelection`](https://github.com/vaadin/flow-components/pull/2954/files#diff-ccb763b854e892ae3aa6c588f4098769bb35a60db672882ac9dfb852389c6dadR428).

Reply:
Yeah, might be easier to comprehend, even though the condition would always fail anyways. Added.